master/buildbot/test/unit/test_process_properties.py

```diff
@@ -408,6 +408,29 @@ def test_property_colon_minus(self):
                       "echo buildby-blddef")
         return d
 
+    def test_notHasKey_changed(self):
+        # let's change the _notHasKey constant after Interpolate instantiation
+        # see http://trac.buildbot.net/ticket/3505
+        # this can happen after some reloads (reconfig?)
+        # can't demonstrate by simply reloading 'properties' from here, though
+        self.props.setProperty("buildername", "linux4", "test")
+        command = Interpolate("echo buildby-%(prop:buildername:-blddef)s")
+
+        from buildbot.process import properties
+        saved_notHasKey = properties._notHasKey
+
+        def restore(result):
+            """Avoid making other tests mad"""
+            properties._notHasKey = saved_notHasKey
+            return result
```

Comment:
there is self.patch to do patch + restore pattern.
http://twistedmatrix.com/documents/15.0.0/api/twisted.trial.unittest.SynchronousTestCase.html#patch